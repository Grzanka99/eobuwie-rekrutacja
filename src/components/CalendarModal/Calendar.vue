<script lang="ts" setup>
import * as Calendar from "@cgsh/calendarts";
import { Day } from "@cgsh/calendarts/build/main/types";
import { DateTime } from "luxon";
import { computed, ref } from "vue";
import { IPickedDates, DaysOfWeek } from "~/types";

const props = defineProps<{
  isOpen: boolean;
  reservedDates: IPickedDates[];
}>();

const currentMonth = ref(
  Calendar.getMonth(DateTime.now().month - 1, DateTime.now().year)
);

const currentMonthArray = computed<Day[][]>(
  () => Calendar.getMonthArray(currentMonth.value) as unknown as Day[][]
);

const isReserved = (day: Day): boolean =>
  !!props.reservedDates.filter(
    (reservedDate) =>
      day.fullJSONDate >= reservedDate.startDate.fullJSONDate &&
      day.fullJSONDate <= reservedDate.endDate.fullJSONDate
  ).length;

const picked = ref<IPickedDates>({
  startDate: null,
  endDate: null,
});

const pickedStatus = (day: Day): 0 | 1 | 2 | 3 | 4 => {
  if (!picked.value.startDate && !picked.value.endDate) return 0;
  const startDate = picked.value.startDate.fullJSONDate;
  const endDate = picked.value.endDate.fullJSONDate;
  const toCheck = day.fullJSONDate;

  if (toCheck === startDate && startDate === endDate) return 4;
  if (toCheck === startDate) return 1;
  if (toCheck === endDate) return 2;
  if (toCheck > startDate && toCheck < endDate) return 3;

  return 0;
};

const counter = ref(0);

const validatePicked = (val: IPickedDates) => {
  if (val.startDate.fullJSONDate > val.endDate.fullJSONDate) {
    const temp = val.startDate;
    val.startDate = val.endDate;
    val.endDate = temp;
    counter.value = 1;
  }

  return true;
};

const pickMe = (day: Day) => {
  if (isReserved(day)) return;

  if (counter.value === 0) {
    picked.value.startDate = day;
    if (!picked.value.endDate) picked.value.endDate = day;
    counter.value = 1;
  } else if (counter.value === 1) {
    picked.value.endDate = day;
    counter.value = 0;
  }

  validatePicked(picked.value);
};
</script>

<template>
  <div v-if="isOpen" class="calendar">
    <ol class="calendar__container">
      <div class="calendar__week">
        <li v-for="name in DaysOfWeek" :key="name" class="flex center">
          {{ name }}
        </li>
      </div>
      <div
        class="calendar__week"
        v-for="(week, i) in currentMonthArray"
        :key="i"
      >
        <li
          v-for="day in week"
          :key="day.fullJSONDate"
          class="calendar__week__day"
          :class="{ virtual: !day.isInCurrentMonth, reserved: isReserved(day) }"
          :data-picked="pickedStatus(day)"
          @click="pickMe(day)"
        >
          {{ day.number }}
        </li>
      </div>
    </ol>
  </div>
</template>

<style lang="scss" scoped>
@import "../../assets/scss/font.scss";

.calendar {
  width: 300px;
  height: 380px;
  background-color: var(--background);
  border-radius: var(--border-radius-small);
  box-shadow: 0px 10px 40px 0px rgba(0, 0, 0, 0.15);

  &__week {
    display: grid;
    grid-template-columns: repeat(7, 1fr);

    &__day {
      @include montserrat();

      cursor: pointer;
      position: relative;
      width: 100%;
      z-index: 1;

      &.virtual,
      &.reserved {
        opacity: 0.5;
      }

      &.reserved {
        cursor: not-allowed;
      }

      &[data-picked="1"],
      &[data-picked="2"],
      &[data-picked="3"],
      &[data-picked="4"] {
        background-color: var(--background-green-light);
        color: var(--background);
      }

      &[data-picked="1"] {
        border-top-left-radius: var(--border-radius-small);
        border-bottom-left-radius: var(--border-radius-small);
      }

      &[data-picked="2"] {
        border-top-right-radius: var(--border-radius-small);
        border-bottom-right-radius: var(--border-radius-small);
      }

      &[data-picked="3"] {
        color: var(--color-primary);
      }

      &[data-picked="4"] {
        background-color: transparent;
      }

      &[data-picked="1"],
      &[data-picked="2"],
      &[data-picked="4"] {
        &::after {
          content: "";
          position: absolute;
          top: 0;
          left: 0;
          width: 100%;
          height: 100%;
          background-color: var(--color-primary);
          border-radius: var(--border-radius-small);
          z-index: -1;
        }
      }
    }
  }

  &__container {
    display: flex;
    flex-direction: column;
    width: 100%;
    gap: var(--padding-tiny);

    list-style: none;
    padding: var(--padding-large);
    margin: 0;
    font-size: var(--font-size-small);

    li {
      display: flex;
      justify-content: center;
      align-items: center;

      padding: 0;
      margin: 0;

      height: 36px;
    }
  }
}
</style>
