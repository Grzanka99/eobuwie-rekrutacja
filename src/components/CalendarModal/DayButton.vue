<script lang="ts" setup>
import { DateTime } from "luxon";
import { Day } from "@cgsh/calendarts/build/main/types";
import { IPickedDates } from "../../types";
import { useCalendarHelpers } from "~/composables/calendar.composable";

const { isReserved } = useCalendarHelpers();

const props = defineProps<{
  day: Day;
  reservedDates: IPickedDates[];
  picked: IPickedDates;
  onPick: (day: Day) => void;
}>();

const pickedStatus = (day: Day): 0 | 1 | 2 | 3 | 4 => {
  if (!props.picked.startDate || !props.picked.endDate) return 0;
  const startDate = props.picked.startDate.fullJSONDate;
  const endDate = props.picked.endDate.fullJSONDate;
  const toCheck = day.fullJSONDate;

  if (toCheck === startDate && startDate === endDate) return 4;
  if (toCheck === startDate) return 1;
  if (toCheck === endDate) return 2;
  if (toCheck > startDate && toCheck < endDate) return 3;

  return 0;
};
</script>

<template>
  <li
    class="day-button"
    :class="{
      virtual: !day.isInCurrentMonth,
      reserved: isReserved(day, props.reservedDates),
      today: DateTime.now().toJSON() === day.fullJSONDate,
    }"
    :data-picked="pickedStatus(day)"
    @click="onPick(day)"
  >
    {{ day.number }}
  </li>
</template>

<style lang="scss" scoped>
.day-button {
  cursor: pointer;
  position: relative;
  width: 100%;
  z-index: 1;

  &.virtual,
  &.reserved {
    opacity: 0.5;
  }

  &.reserved {
    cursor: not-allowed;
  }

  &[data-picked="1"],
  &[data-picked="2"],
  &[data-picked="3"],
  &[data-picked="4"] {
    background-color: var(--background-green-light);
    color: var(--background);
  }

  &[data-picked="1"] {
    border-top-left-radius: var(--border-radius-small);
    border-bottom-left-radius: var(--border-radius-small);
  }

  &[data-picked="2"] {
    border-top-right-radius: var(--border-radius-small);
    border-bottom-right-radius: var(--border-radius-small);
  }

  &[data-picked="3"] {
    color: var(--color-primary);
  }

  &[data-picked="4"] {
    background-color: transparent;
  }

  &[data-picked="1"],
  &[data-picked="2"],
  &[data-picked="4"] {
    &::after {
      content: "";
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      background-color: var(--color-primary);
      border-radius: var(--border-radius-small);
      z-index: -1;
    }
  }
}
</style>
