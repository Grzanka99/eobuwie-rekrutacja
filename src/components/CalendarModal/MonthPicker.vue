<script lang="ts" setup>
import { Months } from "~/types";
import ArrowIcon from "~/components/svg/arrow.svg.vue";

defineProps<{
  prev: () => void;
  next: () => void;
  currentMonth: number;
}>();
</script>

<template>
  <div class="switch-month">
    <button @click="prev" class="prev">
      <arrow-icon />
    </button>
    {{ Months[currentMonth] }}
    <button @click="next" class="next">
      <arrow-icon rev />
    </button>
  </div>
</template>

<style lang="scss" scoped>
.switch-month {
  display: flex;
  align-items: center;
  justify-content: space-between;

  width: 100%;
  border: 1px solid var(--text-grey);
  border-radius: var(--border-radius);
  height: var(--base-input-height);
  padding: 0 var(--padding-smaller);

  & > button {
    border: none;
    background: transparent;
    display: flex;
    justify-content: center;
    align-items: center;

    cursor: pointer;
    transition: transform 75ms ease-in;

    &:hover {
      transform: scale(1.1);
    }
  }
}
</style>
