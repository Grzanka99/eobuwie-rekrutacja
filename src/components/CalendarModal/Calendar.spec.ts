import { mount } from "@vue/test-utils";
import CalendarModal from "./Calendar.vue";

const props = {
  isOpen: true,
  reservedDates: [],
  toggleModal: jest.fn(),
  picked: {
    startDate: null,
    endDate: null,
  },
};

describe("CalendarModal", () => {
  it("should render CalendarModal", () => {
    const wrapper = mount(CalendarModal, { props });
    expect(wrapper.find(".calendar").exists()).toBe(true);
  });

  it("should not render CalendarModal", () => {
    const wrapper = mount(CalendarModal, {
      props: {
        ...props,
        isOpen: false,
      },
    });

    expect(wrapper.find(".calendar").exists()).toBe(false);
  });

  it("should render as expected", () => {
    const wrapper = mount(CalendarModal, { props });
    expect(wrapper.element).toMatchSnapshot();
  });
});
