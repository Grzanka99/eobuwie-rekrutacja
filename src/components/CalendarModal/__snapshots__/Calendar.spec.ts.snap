// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarModal should render as expected 1`] = `
<div
  class="calendar"
>
  <div
    class="switch-month"
  >
    <button
      class="prev"
    >
      <svg
        class=""
        fill="none"
        height="14"
        viewBox="0 0 8 14"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.7 13.405C6.60134 13.4063 6.50344 13.3875 6.41231 13.3497C6.32117 13.3118 6.23872 13.2558 6.17 13.185L0.955 7.97C0.698525 7.71232 0.554539 7.36356 0.554539 7C0.554539 6.63644 0.698525 6.28767 0.955 6.03L6.17 0.814997C6.23929 0.744852 6.32182 0.68916 6.4128 0.65115C6.50378 0.613139 6.6014 0.593567 6.7 0.593567C6.7986 0.593567 6.89622 0.613139 6.9872 0.65115C7.07818 0.68916 7.16071 0.744852 7.23 0.814997C7.36924 0.956261 7.4473 1.14665 7.4473 1.345C7.4473 1.54335 7.36924 1.73373 7.23 1.875L2.105 7L7.23 12.125C7.36924 12.2663 7.4473 12.4566 7.4473 12.655C7.4473 12.8533 7.36924 13.0437 7.23 13.185C7.16093 13.2553 7.07841 13.3111 6.98736 13.3489C6.89631 13.3867 6.79858 13.4057 6.7 13.405Z"
          fill="#333333"
        />
      </svg>
    </button>
     April 
    <button
      class="next"
    >
      <svg
        class="arrowIconRev"
        fill="none"
        height="14"
        viewBox="0 0 8 14"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.7 13.405C6.60134 13.4063 6.50344 13.3875 6.41231 13.3497C6.32117 13.3118 6.23872 13.2558 6.17 13.185L0.955 7.97C0.698525 7.71232 0.554539 7.36356 0.554539 7C0.554539 6.63644 0.698525 6.28767 0.955 6.03L6.17 0.814997C6.23929 0.744852 6.32182 0.68916 6.4128 0.65115C6.50378 0.613139 6.6014 0.593567 6.7 0.593567C6.7986 0.593567 6.89622 0.613139 6.9872 0.65115C7.07818 0.68916 7.16071 0.744852 7.23 0.814997C7.36924 0.956261 7.4473 1.14665 7.4473 1.345C7.4473 1.54335 7.36924 1.73373 7.23 1.875L2.105 7L7.23 12.125C7.36924 12.2663 7.4473 12.4566 7.4473 12.655C7.4473 12.8533 7.36924 13.0437 7.23 13.185C7.16093 13.2553 7.07841 13.3111 6.98736 13.3489C6.89631 13.3867 6.79858 13.4057 6.7 13.405Z"
          fill="#333333"
        />
      </svg>
    </button>
  </div>
  <ol
    class="calendar__container"
  >
    <div
      class="calendar__week"
    >
      
      <li
        class="flex center"
      >
        Mon
      </li>
      <li
        class="flex center"
      >
        Tue
      </li>
      <li
        class="flex center"
      >
        Wed
      </li>
      <li
        class="flex center"
      >
        Thu
      </li>
      <li
        class="flex center"
      >
        Fri
      </li>
      <li
        class="flex center"
      >
        Sat
      </li>
      <li
        class="flex center"
      >
        Sun
      </li>
      
    </div>
    
    <div
      class="calendar__week"
    >
      
      <li
        class="day-button virtual"
        data-picked="0"
      >
        28
      </li>
      <li
        class="day-button virtual"
        data-picked="0"
      >
        29
      </li>
      <li
        class="day-button virtual"
        data-picked="0"
      >
        30
      </li>
      <li
        class="day-button virtual"
        data-picked="0"
      >
        31
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        1
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        2
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        3
      </li>
      
    </div>
    <div
      class="calendar__week"
    >
      
      <li
        class="day-button"
        data-picked="0"
      >
        4
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        5
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        6
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        7
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        8
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        9
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        10
      </li>
      
    </div>
    <div
      class="calendar__week"
    >
      
      <li
        class="day-button"
        data-picked="0"
      >
        11
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        12
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        13
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        14
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        15
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        16
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        17
      </li>
      
    </div>
    <div
      class="calendar__week"
    >
      
      <li
        class="day-button"
        data-picked="0"
      >
        18
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        19
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        20
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        21
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        22
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        23
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        24
      </li>
      
    </div>
    <div
      class="calendar__week"
    >
      
      <li
        class="day-button"
        data-picked="0"
      >
        25
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        26
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        27
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        28
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        29
      </li>
      <li
        class="day-button"
        data-picked="0"
      >
        30
      </li>
      <li
        class="day-button virtual"
        data-picked="0"
      >
        1
      </li>
      
    </div>
    
  </ol>
</div>
`;
